src/build_helper/cc.rs

```diff
@@ -0,0 +1,496 @@
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Invoke the appropriate toolchain to build and assemble a static
+//! library. Different flavours of the toolchain (gnu vs. msvc) are
+//! abstracted behind the trait `Toolchain`, and the build system
+//! will select the appropriate one based on the build configuration
+//! and the platform.
+
+use std::fmt;
+use std::ffi::{OsStr, OsString};
+use std::path::{PathBuf, Path};
+use std::process::Command;
+use llvm::LLVMTools;
+use config::Config;
+use run::Run;
+
+/// Specify a target triple, in the format `arch-vendor-os-abi`.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Triple {
+    triple : String
+}
+
+impl Triple {
+    pub fn new(triple : &str) -> Result<Triple, String> {
+        let v : Vec<&str>= triple.split('-').map(|s| s).collect();
+        if v.len() < 3 {
+            Err(format!("Invalid target triple {}.", triple))
+        } else {
+            Ok(Triple { triple : triple.into() })
+        }
+    }
+
+    pub fn arch(&self) -> &str {
+        self.triple.split('-').nth(0).unwrap()
+    }
+
+    pub fn os(&self) -> &str {
+        self.triple.split('-').nth(2).unwrap()
+    }
+
+    pub fn abi(&self) -> Option<&str> {
+        self.triple.split('-').nth(3)
+    }
+
+    pub fn is_i686(&self) -> bool {
+        self.arch() == "i686"
+    }
+
+    pub fn is_x86_64(&self) -> bool {
+        self.arch() == "x86_64"
+    }
+
+    pub fn is_aarch64(&self) -> bool {
+        self.arch() == "aarch64"
+    }
+
+    pub fn is_windows(&self) -> bool {
+        self.os() == "windows"
+    }
+
+    pub fn is_mingw(&self) -> bool {
+        self.is_windows() && self.abi() == Some("gnu")
+    }
+
+    pub fn is_msvc(&self) -> bool {
+        self.abi() == Some("msvc")
+    }
+
+    pub fn is_linux(&self) -> bool {
+        self.os() == "linux"
+    }
+
+    pub fn is_ios(&self) -> bool {
+        self.os() == "ios"
+    }
+
+    pub fn is_android(&self) -> bool {
+        self.os() == "android"
+    }
+
+    pub fn is_darwin(&self) -> bool {
+        self.os() == "darwin"
+    }
+
+    /// Append the extension for the executables in this platform
+    pub fn with_exe_ext(&self, name : &str) -> String {
+        if self.is_windows() {
+            format!("{}.exe", name)
+        } else {
+            format!("{}", name)
+        }
+    }
+
+    /// Append the extension for the static libraries in this platform
+    pub fn with_lib_ext(&self, name : &str) -> String {
+        if self.is_msvc() {
+            format!("{}.lib", name)
+        } else {
+            format!("lib{}.a", name)
+        }
+    }
+
+    /// Get the file extension for the dynamic libraries in this platform.
+    pub fn dylib_ext(&self) -> &'static str {
+        if self.is_windows() {
+            "dll"
+        } else if self.is_darwin() {
+            "dylib"
+        } else {
+            "so"
+        }
+    }
+}
```

Comment:
It looks like some of these functions are never used, and other have relatively trivial implementations. I'm a little worried about having this as an "official target triple" class which is basically duplication with what we have in the compiler, so I'd be fine just looking inside strings and having various other helper functions. (e.g. `struct Triple` may be overkill)